drafts/bitcartel-nu0-network-handshake/draft.rst

```diff
@@ -95,36 +95,62 @@ Network Coalescence
 
-This can be activated at n blocks before the activation block height, where n could be defined by a constant such as::
+Currently, an eviction process takes place when new inbound connections arrive, but the node has already connected to the maximum number of inbound peers::
 
-    static const int NETWORK_COALESCE_BLOCK_PERIOD = 1000.
+    if (nInbound >= nMaxInbound)
+    {
+        if (!AttemptToEvictConnection(whitelisted)) {
+            // No connection to evict, disconnect the new connection
+            LogPrint("net", "failed to find an eviction candidate - connection dropped (full)\n");
+            CloseSocket(hSocket);
+            return;
+        }
+    }
+
+We update this process by adding behaviour so that the set of eviction candidates will prefer pre-Overwinter node, when the chain tip is in a period N blocks before the activation block height, where N is defined as::
+
+    static const int NETWORK_UPGRADE_PEER_PREFERENCE_BLOCK_PERIOD = 1000.
 
-The eviction code can be updated as follows::
+The eviction candidates can be modified as so::
 
     static bool AttemptToEvictConnection(bool fPreferNewConnection) {
     ...
     // Protect connections with certain characteristics
     ...
-    // Check version of eviction candidates
-    if (current_block_height >= (activationheight - NETWORK_COALESCE_BLOCK_PERIOD)) {
-      // if there exist any legacy nodes, keep them in the eviction set
-      // and at the same time remove overwinter nodes from eviction set.
-      // if there do not exist any legacy nodes,
-      // continue with existing behaviour.
+    // Check version of eviction candidates...
+    // If we are connected to any pre-Overwinter nodes, keep them in the eviction set and remove any Overwinter nodes
+    // If we are only connected to Overwinter nodes, continue with existing behaviour.
+    if ((height < nActivationHeight) &&
+        (height >= (nActivationHeight - NETWORK_UPGRADE_PEER_PREFERENCE_BLOCK_PERIOD)))
+    {
+        // Find any nodes which don't support Overwinter protocol version
+        BOOST_FOREACH(const CNodeRef &node, vEvictionCandidates) {
+            if (node->nVersion < OVERWINTER_PROTO_VERSION) {
+                vTmpEvictionCandidates.push_back(node);
+            }
+        }
 
+        // Prioritize these nodes by replacing eviction set with them
+        if (vTmpEvictionCandidates.size() > 0) {
+            vEvictionCandidates = vTmpEvictionCandidates;
+        }
+    }
 
-Disconnecting Existing Connections
-----------------------------------
+The existing method of disconnecting a candidate remains:
+
+    vEvictionCandidates[0]->fDisconnect = true;
+
+The existing eviction process also classifies and divides eviction candidates into groups.  If a group only has one peer, it will not be evicted.  This means at least one pre-Overwinter node should remain connected upto the activation block height, barring any network issues or a high ban score.
```

Comment:
If  you want to keep a preoverwinter node, shouldn't the condition in the `if` be `>1` rather than `>0` ?

Reply:
The '>0' is for switching out the eviction candidates (which may be a mix of pre-Overwinter and Overwinternodes) when we find 1 or more pre-Overwinter nodes. That way, the set of eviction candidates is only pre-Overwinter.

The sentence above refers to upstream code around 'netgroups' which is later on in the method.

Reply:
I'm wondering if you'll switch a group of more than one candidates, to only 1 pre-overwinter candidates, and then fail to evict cause you only have one.